java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java

```diff
@@ -0,0 +1,290 @@
+    }
+    return wildcard;
+  }
+
+  private JavaType applySubstitution(JavaType.ArrayJavaType arrayType, TypeSubstitution substitution) {
+    JavaType rootElementType = arrayType.elementType;
+    int nbDimensions = 1;
+    while (rootElementType.isArray()) {
+      rootElementType = ((JavaType.ArrayJavaType) rootElementType).elementType;
+      nbDimensions++;
+    }
+    JavaType substitutedType = substitution.substitutedType(rootElementType);
+    if (substitutedType != null) {
+      // FIXME SONARJAVA-1574 a new array type should not be created but reused if already existing for the current element type
+      for (int i = 0; i < nbDimensions; i++) {
+        substitutedType = new JavaType.ArrayJavaType(substitutedType, symbols.arrayClass);
+      }
+      return substitutedType;
+    }
+    return arrayType;
+  }
+
+  TypeSubstitution getSubstitutionFromTypeParams(List<JavaType.TypeVariableJavaType> typeVariableTypes, List<JavaType> typeParams) {
+    TypeSubstitution substitution = new TypeSubstitution();
+    if (typeVariableTypes.size() == typeParams.size()) {
+      // create naive substitution
+      for (int i = 0; i < typeVariableTypes.size(); i++) {
+        JavaType.TypeVariableJavaType typeVariableType = typeVariableTypes.get(i);
+        JavaType typeParam = typeParams.get(i);
+        substitution.add(typeVariableType, typeParam);
+      }
+    }
+    return substitution;
+  }
+
+  private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol method, List<JavaType> formals, List<JavaType> argTypes) {
+    TypeSubstitution substitution = new TypeSubstitution();
+
+    if (formals.size() > argTypes.size() || formals.isEmpty() || argTypes.isEmpty()) {
+      // no need to try to infer types
+      return substitution;
+    }
+
+    List<JavaType.TypeVariableJavaType> typeVariablesToInfer = new ArrayList<>(method.typeVariableTypes);
+    // compete with type variable from the scope
+    for (JavaType argType : argTypes) {
+      JavaSymbol owner = argType.symbol.owner;
+      if (owner.isMethodSymbol() && ((JavaSymbol.MethodJavaSymbol) owner).typeVariableTypes.contains(argType)) {
+        typeVariablesToInfer.add((JavaType.TypeVariableJavaType) argType);
+      }
+    }
+
+    if (!typeVariablesToInfer.isEmpty()) {
+      for (int i = 0; i < formals.size(); i++) {
+        JavaType formalType = formals.get(i);
+        JavaType argType = argTypes.get(i);
+        boolean isLastParam = i == formals.size() - 1;
+
+        substitution = inferTypeSubstitution(method, substitution, isLastParam, formalType, argType);
+
+        if (substitution.typeVariables().containsAll(method.typeVariableTypes)) {
+          // we found all the substitution
+          break;
+        }
+      }
+    }
+    return substitution;
+  }
+
+  private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol method, TypeSubstitution currentSubstitution, boolean isLastParam, JavaType formalType,
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "inferTypeSubstitution" is 22 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)